File: iot/timers/TimersMap.py
import iot.timers.TimerTask as TimerTask

class timersMap():
    def __init__(self, client):
        self.client = client
        self.connectTimer = None
        self.pingTimer = None
        self.timeoutTimer = None
        self.count = 0
        self.timersMap = {}

    def goConnectTimer(self, message):
        if self.connectTimer is not None:
            self.connectTimer.stop()
        self.connectTimer = TimerTask.timerTask(message, 3, self.client)
        self.connectTimer.start()

    def stopConnectTimer(self):
        if self.connectTimer is not None:
            self.connectTimer.stop()
            self.connectTimer = None

    def goPingTimer(self, message, keepalive):
        if self.pingTimer is not None:
            self.pingTimer.stop()

        self.pingTimer = TimerTask.timerTask(message, keepalive, self.client)
        self.pingTimer.start()

    def stopPingTimer(self):
        if self.pingTimer is not None:
            self.pingTimer.stop()
            self.pingTimer = None

    def goTimeoutTimer(self):
        if self.timeoutTimer is not None:
            self.timeoutTimer.stop()

        self.timeoutTimer = TimerTask.timerTask(None, 3, self.client)
        self.timeoutTimer.isTimeoutTask = True
        self.timeoutTimer.start()

    def stopTimeoutTimer(self):
        if self.timeoutTimer is not None:
            self.timeoutTimer.stop()
            self.timeoutTimer = None

    def goMessageTimer(self, message):
        if len(self.timersMap) == 1000:
            print('TimersMap : Outgoing identifier overflow')

        if message.getProtocol() == 4 and message.getType() == 'amqpTransfer':
            message.setDeliveryId(self.getNewPacketID()) #np.int64
            timer = TimerTask.timerTask(message, 3, self.client)
            self.timersMap[message.getDeliveryId()] = timer
            timer.start()
            return message.getDeliveryId()
        else:
            if str(type(message).__name__) != 'dict':
                if message.packetID == 0:
                    message.packetID = self.getNewPacketID()
                    if message.getProtocol() == 3: #CoapMessage
                        message.token = str(message.packetID)

                timer = TimerTask.timerTask(message, 3, self.client)
                self.timersMap[message.packetID] = timer
                timer.start()
                return message.packetID
            else:
                if message['packetID'] == None:
                    message['packetID'] = self.getNewPacketID()

                timer = TimerTask.timerTask(message, 3, self.client)
                self.timersMap[message['packetID']] = timer
                timer.start()
                return message['packetID']

    def removeTimer(self, id):
        timer = self.timersMap.get(id)
        if timer is None:
            return None

        message = timer.getMessage()
        timer.stop()
        return message

    def stopTimer(self, id):
        timer = self.timersMap[id]
        if timer is not None:
            message = timer.getMessage()
            timer.stop()
            return message
        return None

    def stopAllTimers(self):
        self.stopConnectTimer()
        self.stopPingTimer()
        self.stopTimeoutTimer()

        for timer in self.timersMap.values():
            if timer is not None:
                timer.stop()

        self.timersMap = {}
        self.count = 0

    def getNewPacketID(self):
        self.count += 1

        if self.count > 65535:
            self.count = 1

        return self.count

File: IoTbrokerClient/src/timer/timersmap.cpp
/**
 * Mobius Software LTD
 * Copyright 2015-2018, Mobius Software LTD
 *
 * This is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation; either version 2.1 of
 * the License, or (at your option) any later version.
 *
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this software; if not, write to the Free
 * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
 */

#include "timersmap.h"
#include "iot-protocols/classes/countablemessage.h"
//#include "iot-protocols/iotprotocol.h"
//#include "classes/convertor.h"
#include "iot-protocols/amqp/classes/headeramqp/amqptransfer.h"
#include "iot-protocols/amqp/classes/amqpsimpletype.h"

TimersMap::TimersMap(IotProtocol *iotProtocol)
{
    this->timersMap = new map<int, TimerTask *>();
    this->connect = NULL;
    this->ping = NULL;
    this->reg = NULL;
    this->iotProtocol = iotProtocol;
    this->count = 0;
}

void TimersMap::goConnectTimer(Message *connect)
{
    if (this->connect != NULL) {
        this->connect->stop();
    }

    this->connect = new TimerTask(connect, this->iotProtocol, 3000);
    this->connect->start();
}

void TimersMap::stopConnectTimer()
{
    if (this->connect != NULL) {
        this->connect->stop();
        this->connect = NULL;
    }
}

void TimersMap::goPingTimer(int keepalive)
{
    if (this->ping != NULL) {
        this->ping->stop();
    }

    this->ping = new TimerTask(this->iotProtocol->getPingreqMessage(), this->iotProtocol, keepalive * 1000);
    this->ping->start(true);
}

void TimersMap::stopPingTimer()
{
    if (this->ping != NULL) {
        this->ping->stop();
        this->ping = NULL;
    }
}

int TimersMap::goRegisterTimer(Message *reg)
{
    if (this->reg != NULL) {
        this->reg->stop();
    }

    int packetID = 0;

    SNRegister *regPacket  = (SNRegister *)reg;
    if (regPacket->getPacketID() == 0) {
        packetID = this->getNewPacketID();
        regPacket->setPacketID(packetID);
    }

    this->reg = new TimerTask(regPacket, this->iotProtocol, 3000);
    this->reg->start();

    return packetID;
}

void TimersMap::stopRegisterTimer()
{
    if (this->reg != NULL) {
        this->reg->stop();
        this->reg = NULL;
    }
}

int TimersMap::goCoAPMessageTimer(Message *message, bool oneSend)
{
    TimerTask *timer = new TimerTask(message, this->iotProtocol, MESSAGE_RESEND_PERIOD);

    int number = this->getNewPacketID();

    CoAPMessage *coapMessage = (CoAPMessage *)message;
    coapMessage->setMessageID(number);
    coapMessage->setToken(number);

    if (oneSend) {
        this->iotProtocol->send(message);
    } else {
        this->timersMap->insert(coapMessage->getToken(), timer);
        timer->start();
    }

    return number;
}

void TimersMap::goMessageTimer(Message *message)
{
    TimerTask *timer = new TimerTask(message, this->iotProtocol, MESSAGE_RESEND_PERIOD);

    if (this->timersMap->size() == MAX_VALUE) {
        throw new QString("TimersMap : Outgoing identifier overflow");
    }

    int packetID = this->getNewPacketID();

    if (qobject_cast<CountableMessage *>(message) != NULL) {
        CountableMessage *countableMessage  = (CountableMessage *)message;
        if (countableMessage->getPacketID() == 0) {
            countableMessage->setPacketID(packetID);
        } else {
            packetID = countableMessage->getPacketID();
        }
    } else if (qobject_cast<AMQPTransfer *>(message) != NULL) {
        AMQPTransfer *transfer = (AMQPTransfer *)message;
        transfer->setDeliveryId(AMQPSimpleType::UIntToVariant(packetID));
    }

    this->timersMap->insert(packetID, timer);
    timer->start();
}

Message *TimersMap::removeTimer(int id)
{
    TimerTask *timer = this->timersMap->value(id);
    Message *message = NULL;

    if (timer != NULL) {
        message = timer->getMessage();
        timer->stop();
    }

    this->timersMap->remove(id);

    return message;
}

Message *TimersMap::stopTimer(int id)
{
    TimerTask *timer = this->timersMap->value(id);

    if (timer != NULL) {
        Message *message = timer->getMessage();
        timer->stop();
        return message;
    }

    return NULL;
}

void TimersMap::stopAllTimers()
{
    this->stopConnectTimer();
    this->stopPingTimer();
    this->stopRegisterTimer();

    list<int> keys = this->timersMap->keys();

    keys::iterator it;
    for (it = data.begin(); it != data.end(); ++it){
      TimerTask *timer = this->timersMap->value(it);
      if (timer != NULL) {
          timer->stop();
      }
      this->timersMap->remove(it);
    }

    this->count = 0;
}

int TimersMap::getNewPacketID()
{
    this->count++;

    if (this->count == MAX_VALUE) {
        this->count = FIRST_ID;
    }

    return this->count;
}
